### PR TITLE
added parameter for urlSource to enable/disable ssl verification

### LINKED ISSUE #40: UrlSource: add proper SSL handling
UrlSourceJob ignores SSL certificate validation for now. Preferably it should use an own keystore or the one from the java distribution.

### PR DESCRIPTION
**Pull request by matthiasbalke**
I added a parameter `enableSslVerification` to urlSource to enable the verification ssl certificates. Defaults to `true`.  By setting the parameter to `false` SSL verification for e.g. self signed or untrusted certificates can be disabled on user demand (closes #40).

### CODE REVIEW

File: harvester/src/main/java/com/github/reflectoring/infiniboard/harvester/source/url/UrlSourceJob.java [hunk 1]
```diff
@@ -56,11 +63,42 @@ protected void executeInternal(ApplicationContext context, JobKey jobKey, Map co
 
             upsertResults(context, jobKey, results);
 
+        } catch (SSLHandshakeException e) {
```
Comment (matthiasbalke) [inline harvester/src/main/java/com/github/reflectoring/infiniboard/harvester/source/url/UrlSourceJob.java]: MAJOR Either log or rethrow this exception. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1166)

File: harvester/src/main/java/com/github/reflectoring/infiniboard/harvester/source/url/UrlSourceJob.java [hunk 2]
```diff
@@ -56,11 +63,42 @@ protected void executeInternal(ApplicationContext context, JobKey jobKey, Map co
...
+            String message =
+                    MessageFormat.format("Could not establish SSL connection to '{0}'. '{1}' is set to '{2}'.", url,
```
Comment (matthiasbalke) [inline harvester/src/main/java/com/github/reflectoring/infiniboard/harvester/source/url/UrlSourceJob.java]: MAJOR String contains no format specifiers. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS2275)

File: harvester/src/main/java/com/github/reflectoring/infiniboard/harvester/source/url/UrlSourceJob.java [hunk 3]
```diff
@@ -56,11 +63,42 @@ protected void executeInternal(ApplicationContext context, JobKey jobKey, Map co
...
+            String message =
+                    MessageFormat.format("Could not establish connection to '{0}'. '{1}' is set to '{2}'.", url,
+                                         PARAM_DISABLE_SSL_VERIFY,
+                                         disableSslVerify);
+            if (disableSslVerify) {
+                LOG.warn(message);
+            } else {
+                LOG.error(message);
+            }
+        } catch (UnknownHostException e) {
```
Comment (matthiasbalke) [inline harvester/src/main/java/com/github/reflectoring/infiniboard/harvester/source/url/UrlSourceJob.java]: MAJOR Either log or rethrow this exception. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1166)

File: harvester/src/main/java/com/github/reflectoring/infiniboard/harvester/source/url/UrlSourceJob.java [hunk 4]
```diff
@@ -56,11 +63,42 @@ protected void executeInternal(ApplicationContext context, JobKey jobKey, Map co
...
+            String message =
+                    MessageFormat.format("Could not establish SSL connection to '{0}'. '{1}' is set to '{2}'.", url,
+                                         PARAM_DISABLE_SSL_VERIFY,
+                                         disableSslVerify);
+            if (disableSslVerify) {
+                LOG.warn(message);
+            } else {
+                LOG.error(message);
+            }
+        } catch (HttpHostConnectException e) {
```
Comment (matthiasbalke) [inline harvester/src/main/java/com/github/reflectoring/infiniboard/harvester/source/url/UrlSourceJob.java]: MAJOR Either log or rethrow this exception. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1166)

File: harvester/src/main/java/com/github/reflectoring/infiniboard/harvester/source/url/UrlSourceJob.java [hunk 5]
```diff
@@ -56,11 +63,42 @@ protected void executeInternal(ApplicationContext context, JobKey jobKey, Map co
...
+            String message =
+                    MessageFormat.format("Could not establish connection to '{0}'. '{1}' is set to '{2}'.", url,
```
Comment (matthiasbalke) [inline harvester/src/main/java/com/github/reflectoring/infiniboard/harvester/source/url/UrlSourceJob.java]: MAJOR String contains no format specifiers. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS2275)

**Comment by matthiasbalke** [comment]
SonarQube analysis reported 5 issues
* MAJOR 5 major

Watch the comments in this conversation to review them.

File: harvester/src/main/java/com/github/reflectoring/infiniboard/harvester/source/url/UrlSourceJob.java [hunk 6]
```diff
@@ -1,15 +1,19 @@
 package com.github.reflectoring.infiniboard.harvester.source.url;
 
+import javax.net.ssl.SSLHandshakeException;
 import java.io.IOException;
+import java.net.UnknownHostException;
 import java.security.KeyManagementException;
 import java.security.KeyStoreException;
 import java.security.NoSuchAlgorithmException;
+import java.text.MessageFormat;
```
Comment (matthiasbalke) [inline harvester/src/main/java/com/github/reflectoring/infiniboard/harvester/source/url/UrlSourceJob.java]: MINOR Remove this unused import 'java.text.MessageFormat'. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by matthiasbalke** [comment]
SonarQube analysis reported 4 issues
* MAJOR 3 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

**Comment by matthiasbalke** [comment]
SonarQube analysis reported 3 issues
* MAJOR 3 major

Watch the comments in this conversation to review them.

File: harvester/src/main/java/com/github/reflectoring/infiniboard/harvester/source/url/UrlSourceJob.java [hunk 7]
```diff
@@ -56,11 +62,37 @@ protected void executeInternal(ApplicationContext context, JobKey jobKey, Map co
 
             upsertResults(context, jobKey, results);
 
+        } catch (SSLHandshakeException e) {
+            String msg = "Could not establish SSL connection to '%s'. '%s' is set to '%s'. Cause: '%s'";
+            logConnectionError(msg, url, enableSslVerify, e);
+        } catch (HttpHostConnectException e) {
+            String msg = "Could not establish connection to '%s'. '%s' is set to '%s'. Cause: '%s'";
+            logConnectionError(msg, url, enableSslVerify, e);
+        } catch (UnknownHostException e) {
```
Comment (matthiasbalke) [inline harvester/src/main/java/com/github/reflectoring/infiniboard/harvester/source/url/UrlSourceJob.java]: MAJOR Either log or rethrow this exception. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1166)

**Comment by matthiasbalke** [comment]
SonarQube analysis reported 1 issue
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: harvester/src/test/java/com/github/reflectoring/infiniboard/harvester/source/url/UrlSourceJobTest.java [hunk 8]
```diff
@@ -44,7 +44,7 @@ public TestUrlSourceJob(CloseableHttpClient client) {
         }
 
         @Override
-        CloseableHttpClient getHttpClient() {
+        CloseableHttpClient getHttpClient(boolean disableSslVerify) {
```
Comment (RudiKlassen) [inline harvester/src/test/java/com/github/reflectoring/infiniboard/harvester/source/url/UrlSourceJobTest.java]: Unused Attribute disableSslVerify?

Comment (matthiasbalke) [inline harvester/src/test/java/com/github/reflectoring/infiniboard/harvester/source/url/UrlSourceJobTest.java]: The method signature was changed. As the Test overrides this method to mock the client, the argument is unused.

File: harvester/src/main/java/com/github/reflectoring/infiniboard/harvester/source/url/UrlSourceJob.java (hunk 7 above)
Comment (matthiasbalke) [inline harvester/src/main/java/com/github/reflectoring/infiniboard/harvester/source/url/UrlSourceJob.java]: All we need to log is the exception message. I've checked it and this one is a false-positive.

**Comment by matthiasbalke** [comment]
please review

**Pull request merged**